src/storage-types/src/sources.rs

```diff
@@ -183,35 +178,39 @@ impl<S> IngestionDescription<S> {
 
 impl<S: Clone> IngestionDescription<S> {
     pub fn source_exports_with_output_indices(&self) -> BTreeMap<GlobalId, SourceExport<usize, S>> {
```

Comment:
Yes in the future I plan to get rid of the index-based approach and the need to special case the zero-output, but not necessarily by passing through the 'external reference' all the way through. I think more likely is that we will use the `GlobalId` of each `SourceExport` to do the mapping, e.g. we would switch the return type of `SourceRender::render` from `StackedCollection<G, (usize, Result<SourceMessage, DataflowError>)>,` to `StackedCollection<G, (GlobalId, Result<SourceMessage, DataflowError>)>,` 

I think it may be clearer in the next PRs (https://github.com/MaterializeInc/materialize/pull/28676 and https://github.com/MaterializeInc/materialize/pull/28671) why the `external_reference` field doesn't make sense to use as the stream-identifier, because those allow separate `SourceExport`s to ingest the same external reference with their own resume-upper and relations for decoding things, such that they are treated as entirely separate output streams within the collection returned by each source